tests/ebpf_xdp.h

```diff
@@ -76,186 +78,14 @@ typedef unsigned long long u64;
-static int (*bpf_get_current_comm)(void *buf, int buf_size) =
-	(void *) BPF_FUNC_get_current_comm;
-static int (*bpf_perf_event_read)(void *map, int index) =
-	(void *) BPF_FUNC_perf_event_read;
-static int (*bpf_clone_redirect)(void *ctx, int ifindex, int flags) =
-	(void *) BPF_FUNC_clone_redirect;
-static int (*bpf_redirect)(int ifindex, int flags) =
-	(void *) BPF_FUNC_redirect;
-static int (*bpf_perf_event_output)(void *ctx, void *map,
-				    unsigned long long flags, void *data,
-				    int size) =
-	(void *) BPF_FUNC_perf_event_output;
-static int (*bpf_get_stackid)(void *ctx, void *map, int flags) =
-	(void *) BPF_FUNC_get_stackid;
-static int (*bpf_probe_write_user)(void *dst, void *src, int size) =
-	(void *) BPF_FUNC_probe_write_user;
-static int (*bpf_current_task_under_cgroup)(void *map, int index) =
-	(void *) BPF_FUNC_current_task_under_cgroup;
-static int (*bpf_skb_get_tunnel_key)(void *ctx, void *key, int size, int flags) =
-	(void *) BPF_FUNC_skb_get_tunnel_key;
-static int (*bpf_skb_set_tunnel_key)(void *ctx, void *key, int size, int flags) =
-	(void *) BPF_FUNC_skb_set_tunnel_key;
-static int (*bpf_skb_get_tunnel_opt)(void *ctx, void *md, int size) =
-	(void *) BPF_FUNC_skb_get_tunnel_opt;
-static int (*bpf_skb_set_tunnel_opt)(void *ctx, void *md, int size) =
-	(void *) BPF_FUNC_skb_set_tunnel_opt;
-static unsigned long long (*bpf_get_prandom_u32)(void) =
-	(void *) BPF_FUNC_get_prandom_u32;
-static int (*bpf_xdp_adjust_head)(void *ctx, int offset) =
-	(void *) BPF_FUNC_xdp_adjust_head;
-
+/* xdp descriptor similar to sk_buff */
+#define SK_BUFF struct xdp_md
 
-#define load_byte(data, b)  (*(((u8*)(data)) + (b)))
-#define load_half(data, b) __constant_ntohs(*(u16 *)((u8*)(data) + (b)))
-#define load_word(data, b) __constant_ntohl(*(u32 *)((u8*)(data) + (b)))
-#define load_dword(data, b) __constant_ntohl(*(u64 *)((u8*)(data) + (b)))
 #define htonl(d) __constant_htonl(d)
 #define htons(d) __constant_htons(d)
 
-/* a helper structure used by eBPF C program
- * to describe map attributes to elf_bpf loader
- */
-struct bpf_map_def {
-	unsigned int type;
-	unsigned int key_size;
-	unsigned int value_size;
-	unsigned int max_entries;
-    unsigned int map_flags;
-    unsigned int id;
-    unsigned int pinning;
-};
-
-/* used in TC */
-struct bpf_elf_map {
-        __u32 type;
-        __u32 key_size;
-        __u32 value_size;
-        __u32 max_entries;
-        __u32 map_flags;
-        __u32 id;
-        __u32 pinning;
-};
-
-#define REGISTER_START()
-#define REGISTER_TABLE(NAME, TYPE, KEY_SIZE, VALUE_SIZE, MAX_ENTRIES) \
-struct bpf_map_def SEC("maps") NAME = {          \
-    .type       = TYPE,             \
-    .key_size   = KEY_SIZE,         \
-    .value_size = VALUE_SIZE,       \
-    .max_entries    = MAX_ENTRIES,  \
-    .map_flags = 0,                 \
-    .pinning = 2,                   \
-};
-#define REGISTER_END()
+#define load_dword(data, b) (*(u64 *)((u8*)(data) + (b))) << 16
```

Comment:
This is a workaround to make the XDP correctly parse the ethernet src and dst addresses. This is obviously not an ideal solution. I am a little bit confused as to why the kernel and the XDP target behave differently in the u64 case. The difference is that the kernel target only matches keys, but does not write them back. Still, parsing u8, u16, u32 works fine. I guess something is funky the way we parse Ethernet headers right now, but I am not sure about the right way to solve it. Add another function or datatype, try to get rid of the bit shifts?